test/integration.spec.ts

```diff
@@ -1,16 +1,201 @@
 import { TestComponents } from "./test-helpers"
 import FormData = require("form-data")
 import busboy from "busboy"
+import { inspect } from "util"
+import nodeFetch from "node-fetch"
 
 describeE2E("integration sanity tests using express server backend", integrationSuite)
 describeTestE2E("integration sanity tests using test server", integrationSuite)
 
+describeE2E("underlying server", function ({ components }: { components: TestComponents }) {
+  it("gets the underlying http server", async () => {
+    const { server } = components
+    const http = await getUnderlyingServer(server)
+    expect(http.listening).toEqual(true)
+  })
+})
+
+describeTestE2E("underlying server", function ({ components }: { components: TestComponents }) {
+  it("gets the underlying http server", async () => {
+    const { server } = components
+    const http = getUnderlyingServer(server)
+    await expect(http).rejects.toThrow()
+  })
+})
+
 function integrationSuite({ components }: { components: TestComponents }) {
+  it("empty server returns 404", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+
+    const res = await fetch.fetch(`/`)
+    expect(res.status).toEqual(404)
+  })
+
+  it("bypass middleware returns 404", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+    server.use((_, next) => next())
+    const res = await fetch.fetch(`/unexistent-route`)
+    expect(res.status).toEqual(404)
+  })
+
+  it("empty return middleware returns 501", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+    server.use(async () => {
+      return null as any
+    })
+    const res = await fetch.fetch(`/`)
+    expect(res.status).toEqual(501)
+  })
+
+  it("url must use X-Forwarded-Host if available", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+    server.use(async (ctx) => {
+      return { body: ctx.url.toString() }
+    })
+    const res = await fetch.fetch(`/test?a=true`, {
+      headers: { "X-Forwarded-Host": "google.com", host: "arduz.com.ar" },
+    })
+    const url = await res.text()
+    expect(url).toEqual("http://google.com/test?a=true")
+  })
+
+  it("url must use host if available", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+    server.use(async (ctx) => {
+      return { body: ctx.url.toString() }
+    })
+    const res = await fetch.fetch(`/test?a=true`, { headers: { host: "arduz.com.ar" } })
+    const url = await res.text()
+    expect(url).toEqual("http://arduz.com.ar/test?a=true")
+  })
+
+  it("calling multiple next fails", async () => {
+    const { fetch, server } = components
+    server.resetMiddlewares()
+    server.use(async (_, next) => {
+      next()
+      return next()
+    })
+    const res = await fetch.fetch(`/`)
+    expect(res.status).toEqual(500)
+  })
+
+  it("calling multiple next if there is no next returns 404", async () => {
```

Comment:
why is this a 404?

Reply:
imagine this scenario

```ts
routerV1.get('/v1', someHandler)
routerV2.get('/v2', someHandler)

server.use(authMiddleware)
server.use(rateLimitMiddleware)
server.use(routerV1.middleware())
server.use(routerV2.middleware())
```

What would you expect to happen if we call `GET /v1` `GET /v2` and `GET /v3`?

I expect `GET /v1 == 200`, `GET /v2 == 200` and `GET /v3 == 404`

Does it make sense?